amethyst_animation/src/systems/control.rs

```diff
@@ -58,207 +26,186 @@ where
+                    if let AnimationCommand::Step(_) = control.command {
+                        control.command = AnimationCommand::Start;
+                    }
+
+                    if let AnimationCommand::SetInputValue(_) = control.command {
+                        control.command = AnimationCommand::Start;
+                    }
+
+                    // remove completed animations
+                    if remove {
+                        remove_ids.push(*id);
+                    } else {
+                        // record current position of running animations to know when to trigger deferred
+                        state_set.insert(
+                            *id,
+                            match control.state {
+                                ControlState::Running(_) => find_max_duration(
+                                    control.id,
+                                    world
+                                        .entry_ref(
+                                            *hierarchy
+                                                .and_then(|h| h.nodes.values().next())
+                                                .unwrap_or(&entity),
+                                        )
+                                        .unwrap()
+                                        .get_component::<SamplerControlSet<T>>()
+                                        .ok(),
+                                ),
+                                _ => -1.0,
+                            }
+                        );
+                    }
                 }
-                if remove {
-                    self.remove_ids.push(*id);
-                } else {
-                    self.state_set.insert(
-                        *id,
-                        get_running_duration(entity, control, hierarchies.get(entity), &samplers),
-                    );
+
+                // record deferred animation as not started
+                for deferred_animation in &control_set.deferred_animations {
+                    state_set.insert(deferred_animation.animation_id, -1.0);
                 }
-            }
-            for deferred_animation in &control_set.deferred_animations {
-                self.state_set.insert(deferred_animation.animation_id, -1.0);
-            }
-            self.deferred_start.clear();
-            for deferred_animation in &control_set.deferred_animations {
-                let (start, start_dur) =
-                    if let Some(dur) = self.state_set.get(&deferred_animation.relation.0) {
-                        if *dur < 0. {
-                            (false, 0.)
-                        } else if let DeferStartRelation::Start(start_dur) =
-                            deferred_animation.relation.1
-                        {
-                            let remain_dur = dur - start_dur;
-                            (remain_dur >= 0., remain_dur)
+
+                let mut deferred_start = Vec::default();
```

Comment:
This would allocate each time the system is run, whereas the spec version would only clear the already allocated vec. Should we keep the spec behavior? Does something prevent us from doing it the old way?

Reply:
changed back to old behavior